src/raftstore/store/worker/apply.rs

```diff
@@ -1105,6 +1151,270 @@ impl ApplyDelegate {
+        let stash = ctx.stash(self);
+        let mut delegate = match ctx.delegates.get_mut(&source_region.get_id()) {
+            None => panic!("{} source region {:?} not exist", self.tag, source_region),
+            Some(e) => e.take().unwrap_or_else(|| {
+                panic!(
+                    "{} unexpected circle dependency of region {:?}",
+                    self.tag, source_region
+                )
+            }),
+        };
+        delegate.handle_raft_committed_entries(ctx, entries);
+        *ctx.delegates.get_mut(&source_region.get_id()).unwrap() = Some(delegate);
+        ctx.apply_res.last_mut().unwrap().merged = true;
+        ctx.restore_stash(stash);
+    }
+
+    fn exec_commit_merge(
+        &mut self,
+        ctx: &mut ApplyContext,
+        req: &AdminRequest,
+    ) -> Result<(AdminResponse, Option<ExecResult>)> {
+        PEER_ADMIN_CMD_COUNTER_VEC
+            .with_label_values(&["commit_merge", "all"])
+            .inc();
+
+        let merge = req.get_commit_merge();
+        let source_region = merge.get_source();
+        let region_state_key = keys::region_state_key(source_region.get_id());
+        let state: RegionLocalState = match self.engine.get_msg_cf(CF_RAFT, &region_state_key) {
+            Ok(Some(s)) => s,
+            e => panic!(
+                "{} failed to get regions state of {:?}: {:?}",
+                self.tag, source_region, e
+            ),
+        };
+        match state.get_state() {
+            PeerState::Normal | PeerState::Merging => {}
+            _ => panic!(
+                "{} unexpected state of merging region {:?}",
+                self.tag, state
+            ),
+        }
+        let exist_region = state.get_region();
+        if source_region.get_start_key() != exist_region.get_start_key()
+            || source_region.get_end_key() != exist_region.get_end_key()
+        {
+            panic!(
+                "{} source_region {:?} not match exist region {:?}",
+                self.tag, source_region, exist_region
+            );
+        }
+
+        self.catch_up_log_for_merge(ctx, merge, exist_region);
+
+        let mut region = self.region.clone();
+        // Use a max value so that pd can ensure overlapped region has a priority.
+        let version = cmp::max(
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+        write_peer_state(&self.engine, ctx.wb(), &region, PeerState::Normal)
+            .and_then(|_| {
+                // Should source_region be used?
+                write_peer_state(&self.engine, ctx.wb(), exist_region, PeerState::Tombstone)
```

Comment:
can we meet the case that the source region handles applied logs next in the same ready loop?

Reply:
No, it has been check at L2017.